streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -16,276 +16,98 @@
-                    if (sharedTimeTracker.minTime + outerJoinLookBackTimeMs + joinGraceMs >= sharedTimeTracker.streamTime) {
-                        if (timestampedKeyAndJoinSide.isLeftSide()) {
-                            outerJoinLeftBreak = true; // there are no more candidates to emit on left-outerJoin-side
-                        } else {
-                            outerJoinRightBreak = true; // there are no more candidates to emit on right-outerJoin-side
-                        }
-                        if (outerJoinLeftBreak && outerJoinRightBreak) {
-                            break; // there are no more candidates to emit on left-outerJoin-side and
-                                    // right-outerJoin-side
-                        } else {
-                            continue; // there are possibly candidates left on the other outerJoin-side
-                        }
-                    }
-
-                    final VOut nullJoinedValue;
-                    if (isLeftSide) {
-                        nullJoinedValue = joiner.apply(key,
-                                value.getLeftValue(),
-                                value.getRightValue());
-                    } else {
-                        nullJoinedValue = joiner.apply(key,
-                                (V1) value.getRightValue(),
-                                (V2) value.getLeftValue());
-                    }
-
-                    context().forward(
-                        record.withKey(key).withValue(nullJoinedValue).withTimestamp(timestamp)
-                    );
-
-                    if (prevKey != null && !prevKey.equals(timestampedKeyAndJoinSide)) {
-                        // blind-delete the previous key from the outer window store now it is emitted;
-                        // we do this because this delete would remove the whole list of values of the same key,
-                        // and hence if we delete eagerly and then fail, we would miss emitting join results of the later
-                        // values in the list.
-                        // we do not use delete() calls since it would incur extra get()
-                        store.put(prevKey, null);
-                    }
-
-                    prevKey = timestampedKeyAndJoinSide;
-                }
-
-                // at the end of the iteration, we need to delete the last key
-                if (prevKey != null) {
-                    store.put(prevKey, null);
-                }
-            }
-        }
-
-        private long getOuterJoinLookBackTimeMs(final TimestampedKeyAndJoinSide<K> timestampedKeyAndJoinSide) {
+        protected long getOuterJoinLookBackTimeMs(final TimestampedKeyAndJoinSide<K> timestampedKeyAndJoinSide) {
             // depending on the JoinSide we fill in the outerJoinLookBackTimeMs
-            if (timestampedKeyAndJoinSide.isLeftSide()) {
+            if (timestampedKeyAndJoinSide.getJoinSide() == JoinSide.LEFT) {
```

Comment:
A side note: This decision could not be removed. In the KStreamKStreamRightJoin, there are some tests that fetch the value from the rightOuterStore, and the value contains the left-side value.

In the KStreamKStreamLeftJoin this call was not needed.